transformer/cluttered_mnist_rnn.py

```diff
@@ -0,0 +1,219 @@
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# =============================================================================
+import tensorflow as tf
+from spatial_transformer import transformer
+from tensorflow.python.ops import rnn,rnn_cell
+import numpy as np
+from tf_utils import weight_variable, bias_variable, dense_to_one_hot
+
+# %% Load data
+
+
+mnist_cluttered = np.load('./data/mnist_sequence3_sample_8distortions_9x9.npz')
+
+X_train = mnist_cluttered['X_train']
+y_train = mnist_cluttered['y_train']
+X_valid = mnist_cluttered['X_valid']
+y_valid = mnist_cluttered['y_valid']
+X_test = mnist_cluttered['X_test']
+y_test = mnist_cluttered['y_test']
+
+y_train = np.reshape(y_train,[y_train.size,1])
+y_valid = np.reshape(y_valid,[y_valid.size,1])
+y_test = np.reshape(y_test,[y_test.size,1])
+
+# % turn from dense to one hot representation
+Y_train = dense_to_one_hot(y_train, n_classes=10)
+Y_valid = dense_to_one_hot(y_valid, n_classes=10)
+Y_test = dense_to_one_hot(y_test, n_classes=10)
+
+
+# %% Placeholders for 100x100 resolution
+x = tf.placeholder(tf.float32, [None, 10000])
+y = tf.placeholder(tf.float32, [None, 10])
+
+# %% Since x is currently [batch, height*width], we need to reshape to a
+# 4-D tensor to use it in a convolutional graph.  If one component of
+# `shape` is the special value -1, the size of that dimension is
+# computed so that the total size remains constant.  Since we haven't
+# defined the batch dimension's shape yet, we use -1 to denote this
+# dimension should not change size.
+x_tensor = tf.reshape(x, [-1, 100, 100, 1])
+
+#%% localizaton network
+
+keep_prob = tf.placeholder(tf.float32)
+
+
+l_pool0_loc = tf.nn.max_pool(x_tensor,ksize=[1,2,2,1],strides=[1,2,2,1],padding='VALID')
+
+W_conv0_loc = weight_variable([3,3,1,20])
+
+l_conv0_loc = tf.nn.conv2d(l_pool0_loc,W_conv0_loc,strides=[1,1,1,1],padding='VALID')
+
+l_pool1_loc = tf.nn.max_pool(l_conv0_loc,ksize=[1,2,2,1],strides =[1,2,2,1],padding='VALID')
+
+W_conv1_loc = weight_variable([3,3,20,20])                   
+
+l_conv1_loc =  tf.nn.conv2d(l_pool1_loc,W_conv1_loc,strides=[1,1,1,1],padding='VALID')
+
+l_conv1_loc = tf.nn.dropout(l_conv1_loc,keep_prob)
+
+l_pool2_loc = tf.nn.max_pool(l_conv1_loc,ksize=[1,2,2,1],strides =[1,2,2,1],padding='VALID')
+
+W_conv2_loc = weight_variable([3,3,20,20])
+
+l_conv2_loc = tf.nn.conv2d(l_pool2_loc,W_conv2_loc,strides=[1,1,1,1],padding='VALID') 
+
+l_conv2_loc = tf.reshape(l_conv2_loc,[-1 ,9*9*20 ])
+
+# Replicate input for Gated Recurrent Unit
+l_conv2_loc = tf.tile(l_conv2_loc,[1,3])
+
+l_conv2_loc = tf.split(1,3,l_conv2_loc)
+
+# Gated Recurrent Unit
+gru_cell = rnn_cell.GRUCell(num_units=256)
+
```

Comment:
Would you mind removing the empty lines between the model definition? Please follow the style guide.